python/paddle/v2/fluid/distribute_transpiler.py

```diff
@@ -412,30 +422,46 @@ def _append_pserver_ops(self, program, pserver_program, opt_op, endpoint):
                 shape=new_shape)
 
         # change output's ParamOut variable
-        opt_op.outputs["ParamOut"] = new_inputs["Param"]
+        outputs = self._get_output_map_from_op(program.global_block(), opt_op)
+        outputs["ParamOut"] = new_inputs["Param"]
         program.global_block().append_op(
             type=opt_op.type,
             inputs=new_inputs,
-            outputs=opt_op.outputs,
+            outputs=outputs,
             attrs=opt_op.attrs)
 
     def _append_pserver_non_opt_ops(self, program, pserver_program, opt_op):
         # Append the ops for parameters that do not need to be optimized/updated
-        for _, var in opt_op.inputs.iteritems():
-            program.global_block().create_var(
-                name=var.name,
-                persistable=var.persistable,
-                dtype=var.dtype,
-                shape=var.shape)
-            pserver_program.global_block().create_var(
-                name=var.name,
-                persistable=var.persistable,
-                dtype=var.dtype,
-                shape=var.shape)
+        inputs = self._get_input_map_from_op(self.program.global_block().vars,
+                                             opt_op)
+        for var in inputs.itervalues():
+            if type(var) == list:
+                varlist = var
+            else:
+                varlist = [var]
+            for var in varlist:
+                program.global_block().create_var(
+                    name=var.name,
+                    persistable=var.persistable,
+                    dtype=var.dtype,
+                    shape=var.shape)
+                try:
+                    pserver_program.global_block().create_var(
```

Comment:
Can we lookup the var, compare the already created var with the current var, log error if they are not equal?

Reply:
Done and updated. Find var from block, not adding a error log because it's not an error, this will be refined due to https://github.com/PaddlePaddle/Paddle/issues/7700

Reply:
Thanks! Just to double check, wouldn't it be an error if with they are with the same name but the configuration of var (e.g., size) differ?

Reply:
In current implementation, `listen_and_serv` run `optimize_sub_program` with an executor that does not create variable in the sub scope (`create_vars=false`). The variables must only exist in `pserver_program`, just for compile time to pass, we must create variables here. But, once we put the `optimize_sub_program` to the sub-block of `pserver_program` this will not be a problem. Will do this in next PR.

Reply:
Thanks!